Source/CombatExtended/CombatExtended/Verbs/Verb_LaunchProjectileCE.cs

```diff
@@ -348,9 +348,157 @@ public virtual void ShiftTarget(ShiftVecReport report, bool calculateMechanicalO
+                                        targetRange.min = victimVert.BottomHeight;
+                                        targetRange.max = victimVert.MiddleHeight;
+                                    }
+                                }
+                            }
+                        }
+                        else
+                        {
+                            var Victim = (Pawn)CurrentTarget.Thing;
+
+                            targetRange.min = victimVert.BottomHeight;
+                            targetRange.max = victimVert.MiddleHeight;
+
+                            if (((Victim?.kindDef?.RaceProps?.Humanlike ?? false) && AppropiateAimMode))
+                            {
+                                if (ShootingAccuracy >= 2.2f)
+                                {
+                                    #region Finding highest protection apparel on legs, head and torso
+                                    var Head = Victim.health.hediffSet.GetNotMissingParts().Where(X => X.def.defName == "Head").First();
+
+                                    var Helmets = (Victim?.apparel?.WornApparel?.FindAll(F => F.def.apparel.CoversBodyPart(Head)) ?? null);
+
+                                    var Helmet = Helmets.FirstOrDefault();
+
+                                    foreach (Apparel ap in Helmets)
+                                    {
+                                        if (ap.GetStatValue(StatDefOf.ArmorRating_Sharp) > Helmet.GetStatValue(StatDefOf.ArmorRating_Sharp))
+                                        {
+                                            Helmet = ap;
+                                        }
+                                    }
+
+                                    var Leg = Victim.health.hediffSet.GetNotMissingParts().Where(X => X.def.defName == "Leg").First();
+
+                                    var LegArmors = (Victim?.apparel?.WornApparel?.FindAll(F => F.def.apparel.CoversBodyPart(Leg)) ?? null);
+
+                                    var LegArmor = LegArmors.FirstOrDefault();
+
+                                    foreach (Apparel ap in LegArmors)
+                                    {
+                                        if (ap.GetStatValue(StatDefOf.ArmorRating_Sharp) > LegArmor.GetStatValue(StatDefOf.ArmorRating_Sharp))
+                                        {
+                                            LegArmor = ap;
+                                        }
+                                    }
+
+                                    var Torso = Victim.health.hediffSet.GetNotMissingParts().Where(X => X.def.defName == "Torso").First();
+
+                                    var TorsoArmors = (Victim?.apparel?.WornApparel?.FindAll(F => F.def.apparel.CoversBodyPart(Torso)) ?? null);
+
+                                    var TorsoArmor = TorsoArmors.FirstOrDefault();
+
+                                    foreach (Apparel ap in TorsoArmors)

```

Comment:
I suspect a Linq query for the max, using the GetStatValue function as a custom sort function would be faster.  At the very least, cache the armor value of the current-best TorsoArmor, so as to avoid the repeated TorsoArmor.GetStatValue calls (those are expensive).

Same for the other locations.